src/Lecture4.hs

```diff
@@ -184,7 +216,24 @@ instance for the 'Stats' type itself.
 -}
 
 instance Semigroup Stats where
-
+    (<>) :: Stats -> Stats -> Stats
+    a <> b = Stats
+        (statsTotalPositions a <> statsTotalPositions b)
+        (statsTotalSum a <> statsTotalSum b)
+        (statsAbsoluteMax a <> statsAbsoluteMax b)
+        (statsAbsoluteMin a <> statsAbsoluteMin b)
+        (statsSellMax a `combine` statsSellMax b)
+        (statsSellMin a `combine` statsSellMin b)
+        (statsBuyMax a `combine` statsBuyMax b)
+        (statsBuyMin a `combine` statsBuyMin b)
+        (statsLongest a <> statsLongest b)
+
+combine :: Semigroup a => Maybe a -> Maybe a -> Maybe a
+combine x y = case (x, y) of
+    (Just !u, Just !v) -> Just (u <> v)
+    (Just !u, Nothing) -> Just u
+    (Nothing, Just !v) -> Just v
+    (Nothing, Nothing) -> Nothing
```

Comment:
That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by pattern matching only on the final result:

```suggestion
combine x y = case x <> y of
    Nothing -> Nothing
    Just !v -> Just v
```